docs/false_positives.md

```diff
@@ -0,0 +1,243 @@
+targets usually contains assertions. These assertions convey useful information
+to the analyzer.
+
+The analyzer understands the standard assert macro. In case a project has
+custom assertion mechanisms the corresponding functions should be annotated
+both to improve precision and avoid false positives. For details read the
+guide of the [analyzer](https://clang-analyzer.llvm.org/annotations.html#custom_assertions).
+
+
+### <a name="infeasible-path"></a> Infeasible path
+
+One of the most frequent source of false positives is analyzing infeasible 
+execution path of the program. Every report on such a path is false positive.
+There are a number of ways to help the analyzer detect which paths are not
+possible to take. This document also has some examples to show these techniques
+in action.
+
+* Functions that never return should be annotated as noreturn
+* Functions that never return null pointers might be annotated appropriately
+* Assertions can be added to help the analyzer detect more invariants
+* Partial functions can be made more explicit
+* Unreachable annotations should be added
+* Prefer standard functions to custom solutions
+* Do not turn off core checks
+
+#### <a name="correlated-conditions"></a> Correlated conditions
+
+The analyzer might not be able to detect that two conditions are mutually 
+exclusive. Let us consider the following false positive:
+
+```cpp
+int x = f();
+if (condition A) {
+  x = 0;
+}
+...
+if (nontrivial condition) {
+  g(5/x); // Division by zero warning.
+}
+```
+
+It can be rewritten to as the following to suppress the warning:
+
+```cpp
+int x = f();
+if (condition A) {
+  x = 0;
+}
+...
+if (nontrivial condition) {
+  assert(!(condition A));
+  g(5/x); // No warning.
+}
+```
+
+#### <a name="partial-functions"></a> Partial functions
+
+Some functions only work on a set of possible input values. This precondition
+is unknown to the analyzer unless it is expressed explicitly in the code.
+
+Let us consider the following false positive:
+
+```cpp
+int f(MyEnum Val) {
+  int x = 0;
+  switch (Val) {
+    case MyEnumA: x = 1; break;
+    case MyEnumB: x = 5; break;
+  }
+  return 5/x; // Division by zero when Val == MyEnumC.
+}
+```
+
+It can be rewritten to as the following to suppress the warning:
+
+```cpp
+int f(MyEnum Val) {
+  int x = 0;
+  switch (Val) {
+    case MyEnumA: x = 1; break;
+    case MyEnumB: x = 5; break;
+    default: assert(false); break;
+  }
+  return 5/x; // No warning.
+}
+```
+
+Other macros or builtins expressing unreachable code may be used.
+Note that the rewritten code is also safer, since debug builds now check for
+more precondition violations. 
+
+#### <a name="prefer-standard-functions"></a> Prefer standard functions
+
+The analyzer models the behavior of some standard functions but it has
+no knowledge about the semantics of custom made functions declared in a
+library that compiled separately (or in another translation unit).
```

Comment:
A bit of mention about Cross-TU here, perhaps `or in another translation unit if the experimental cross-TU analysis is disabled`? (With link to github.com:Ericsson/clang?)